tun2socks/lwip/device.go

```diff
@@ -0,0 +1,145 @@
+	}
+	lwipLib.RegisterTCPConnHandler(t2s.tcp)
+	lwipLib.RegisterUDPConnHandler(t2s.udp)
+	lwipLib.RegisterOutputFn(t2s.handleIPResponseFromStack)
+
+	return t2s, nil
+}
+
+// Close implements io.Closer and network.IPDevice. It closes the device, rendering it unusable for I/O.
+//
+// Close does not close other objects that are passed to the device, such as StreamDialer, PacketListener, Writer, or
+// Reader. You are responsible for closing these objects yourself.
+func (t2s *LwIPTun2SocksDevice) Close() error {
+	err := t2s.stack.Close()
+	t2s.outPipe.close()
+	if err != nil {
+		return fmt.Errorf("failed to close lwIP stack: %w", err)
+	}
+	return nil
+}
+
+// MTU implements network.IPDevice. It returns the maximum buffer size of a single IP packet that can be processed by
+// this device.
+func (t2s *LwIPTun2SocksDevice) MTU() int {
+	return packetMTU
+}
+
+// lwIP responses from TCP/UDP connections (in IP Packets) callback function
+func (t2s *LwIPTun2SocksDevice) handleIPResponseFromStack(b []byte) (int, error) {
+	return t2s.outPipe.write(b)
+}
+
+// Read implements io.Reader and network.IPDevice. It reads one IP packet from the TCP/UDP response, blocking until a
+// packet arrives or this device is closed. To prevent potential memory allocations, use the WriteTo function.
+func (t2s *LwIPTun2SocksDevice) Read(p []byte) (int, error) {
+	return t2s.outPipe.read(p)
+}
+
+// WriteTo implements io.WriterTo. It writes all IP packets from TCP/UDP responses to `w` until all data is written or
+// an error occurs. This function will not allocate any intermediate buffers.
+//
+// WriteTo returns the total number of bytes written and any error encountered during the write.
+func (t2s *LwIPTun2SocksDevice) WriteTo(w io.Writer) (int64, error) {
+	return t2s.outPipe.writeTo(w)
+}
+
+// Write implements io.Writer and network.IPDevice. It writes a single IP packet to this device. The device will then
+// translate the IP packet into a TCP or UDP traffic.
+func (t2s *LwIPTun2SocksDevice) Write(b []byte) (int, error) {
+	n, err := t2s.stack.Write(b)
+	// workaround: lwip netstack did not use a standard error code
+	if err != nil && err.Error() == "stack closed" {
+		return n, net.ErrClosed
+	}
+	return n, err
+}
+
+// ReadFrom implements io.ReaderFrom. It reads all IP packets from `r` until EOF or error. EOF can occur if either this
+// device is closed or `r` is closed, and it will not be treated as an error. This function will only allocate a single
+// buffer of MTU() bytes.
+//
+// ReadFrom returns the total number of bytes read and any error encountered during the read.
+func (t2s *LwIPTun2SocksDevice) ReadFrom(r io.Reader) (int64, error) {
+	p := make([]byte, t2s.MTU())
+	total := int64(0)
+	for {
+		n, err := r.Read(p)
+		total += int64(n)
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, os.ErrClosed) {
+				break
+			}
+			return total, err
+		}
+		if _, err := t2s.Write(p); err != nil {
+			if errors.Is(err, net.ErrClosed) {
```

Comment:
Better to use `go func() { r.Read(p); rd<-0 }` and then `select { case <-rd: case <-closed: }` to wait for either read or `t2s` is closed. The caller will receive the result immediately though we might read one more unused packet from `r`.

Reply:
It will, but typically the closing sequence is:

```
tun2socks.Close()
// Do sth in between
tun.Close()
```

Only after `tun2socks.Close()`, the read will return an error. Ideally I'd like this function to exit immediately we call `tun2socks.Close()`.

Reply:
See my comments above on the use of channels.

Reply:
I spent some more time thinking about this.

I think a closing action can simply be `tun.Close()`. That aligns with Android, where I think the OS may even close the tun for us.

I imagine the following relay code:
```go
go func() {
  defer t2s.Close()
  for {
    n, err := tun.Read(b)  // Read on a file returns EOF after close
    if err != nil {
     if errors.Is(err, io.EOF) {
       return
     }
      // log error?
     continue
    }
    n, err = t2s.Write(b[:n])
    if err != nil {
      if err is already closed {
        return 
      }
      // log error
    }
  }
}()

go func() {
  defer tun.Close()
  for {
    n, err := t2s.Read(b)  // Should return EOF after close
    if err != nil {
     if errors.Is(err, io.EOF) {
       return
     }
      // log error?
     continue
    }
    n, err = tun.Write(b[:n])
    if err != nil {
      if err is already closed {
        return 
      }
      // log error
    }
  }
}
```

The code is perfectly symmetrical. And closing either tun or tun2socks will break the loop. Calling either one, or even calling both should be allowed.

Reply:
We probably want to standardize on the "already closed" error. There's `net.ErrClosed`, `fs.ErrClosed`, `os.ErrClosed`, `syscall.EPIPE`, ...
Perhaps a function that checks for all of that.

Reply:
Ideally this should work. We must add wrapper around the existing `tun` device because the `tun` implementation in Linux (using `github.com/songgao/water` package) won't return `EOF` but rather an `os.ErrClosed` error to the caller. But anyway, we need to add our wrappers to Windows, Android and iOS.